src/jdk.jpackage/share/classes/jdk/jpackage/internal/util/TokenReplace.java

```diff
@@ -0,0 +1,203 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * Class to replace tokens in strings.
+ * <p>
+ * Single instance holds a list of tokens. Tokens can be substrings of each other.
+ * The implementation performs greedy replacement: longer tokens are replaced first.
+ */
+public final class TokenReplace {
+
+    private record TokenCut(String[] main, String[] sub) {
+        static String[] orderTokens(String... tokens) {
+            if (tokens.length == 0) {
+                throw new IllegalArgumentException("Empty token list");
+            }
+
+            final var orderedTokens = Stream.of(tokens)
+                    .sorted(Comparator.<String>naturalOrder().thenComparing(Comparator.comparingInt(String::length)))
+                    .distinct()
+                    .toArray(String[]::new);
+
+            if (orderedTokens[0].isEmpty()) {
+                throw new IllegalArgumentException("Empty token in the list of tokens");
+            }
+
+            return orderedTokens;
+        }
+
+        static TokenCut createFromOrderedTokens(String... tokens) {
+            final List<Integer> subTokens = new ArrayList<>();
+
+            for (var i = 0; i < tokens.length - 1; ++i) {
+                final var x = tokens[i];
+                for (var j = i + 1; j < tokens.length; ++j) {
+                    final var y = tokens[j];
+                    if (y.contains(x)) {
+                        subTokens.add(i);
+                    }
+                }
+            }
+
+            if (subTokens.isEmpty()) {
+                return new TokenCut(tokens, null);
+            } else {
+                final var main = IntStream.range(0, tokens.length)
+                        .mapToObj(Integer::valueOf)
+                        .filter(Predicate.not(subTokens::contains))
+                        .map(i -> {
+                            return tokens[i];
+                        }).toArray(String[]::new);
+                final var sub = subTokens.stream().map(i -> {
+                    return tokens[i];
+                }).toArray(String[]::new);
+                return new TokenCut(main, sub);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return String.format("TokenCut(main=%s, sub=%s)", Arrays.toString(main), Arrays.toString(sub));
+        }
+    }
+
+    public TokenReplace(String... tokens) {
+        tokens = TokenCut.orderTokens(tokens);
+
+        this.tokens = tokens;
+        regexps = new ArrayList<>();
+
+        for(;;) {
```

Comment:
```suggestion
        for (;;) {
```

Reply:
> If you do echo on macOS in terminal for unset variable it will be expanded to empty string, so for "Unset variables are not expanded." should we have Welcome ! instead of Welcome $USER!?

This is true for any shell behavior. It is also true that some shells fail to expand undefined variables if configured accordingly. Bash. for example:

Reply:
Agree. Makes sense.

Reply:
Oh, I thought I posted my reply in the main thread and then didn't find it there, assumed my comment was lost and wrote it again 🤦